internal/server/providers_test.go

```diff
@@ -120,3 +128,214 @@ func TestAPI_DeleteProvider(t *testing.T) {
 		})
 	}
 }
+
+func TestAPI_CreateProvider(t *testing.T) {
+	srv := setupServer(t, withAdminUser)
+	routes := srv.GenerateRoutes(prometheus.NewRegistry())
+
+	type testCase struct {
+		name     string
+		body     api.CreateProviderRequest
+		setup    func(t *testing.T, req *http.Request)
+		expected func(t *testing.T, response *httptest.ResponseRecorder)
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		body := jsonBody(t, tc.body)
+
+		req, err := http.NewRequest(http.MethodPost, "/api/providers", body)
+		assert.NilError(t, err)
+		req.Header.Add("Authorization", "Bearer "+adminAccessKey(srv))
+		req.Header.Set("Infra-Version", "0.13.6")
+
+		if tc.setup != nil {
+			tc.setup(t, req)
+		}
+
+		resp := httptest.NewRecorder()
+		routes.ServeHTTP(resp, req)
+
+		tc.expected(t, resp)
+	}
+
+	var testCases = []testCase{
+		{
+			name: "not authenticated",
+			setup: func(t *testing.T, req *http.Request) {
+				req.Header.Del("Authorization")
+			},
+			expected: func(t *testing.T, resp *httptest.ResponseRecorder) {
+				assert.Equal(t, resp.Code, http.StatusUnauthorized, resp.Body.String())
+			},
+		},
+		{
+			name: "not authorized",
+			setup: func(t *testing.T, req *http.Request) {
+				accessKey, _ := createAccessKey(t, srv.db, "usera@example.com")
+				req.Header.Set("Authorization", "Bearer "+accessKey)
+			},
+			expected: func(t *testing.T, resp *httptest.ResponseRecorder) {
+				t.Skip("TODO: requires either a fake oidc client, or to authorize before validation")
```

Comment:
The two "not authorized" tests in this PR are skipped right now. 

It's not easy to test this because we validate the request before we authorize it. Part of the validation is checking the `client-id`, and `client-secret` work when sent to the provider at `url`.

We don't want to check-in valid credentials to a provider, so that leaves us two options:

1. change the API handler to authorize before validation. This might be a good thing to do in general, but likely requires replacing the access package (https://github.com/infrahq/internal/issues/20).
2. inject a fake OIDC client so that validation of those fields passes.

We have a lot of the pieces required for 2. There's are a couple `mockOIDCImplementation`, we'd need to export one so that we can use it in other packages. 
Currently the mock is read from `gin.Context`. I think we'll want to change it to read from `context.Context`, so that we can inject it into the request. There's no way to inject a value into the `gin.Context`, which I think is a good thing. We don't want tests relying on `gin.Context` so that we can easily remove it if we need to.

Reply:
We could create another mock client in this package, they're pretty simple. I also like switching things around to authorize before validating, but as you said that's a big change for a small problem.

Reply:
I pushed another commit which uses the fake OIDC client, and these tests are no longer skipped.